src/datasource.ts

```diff
@@ -17,4 +38,157 @@ export class DataSource extends DataSourceWithBackend<RedshiftQuery, RedshiftDat
+      if (!targets.length) {
+        return of({ data: [] });
+      }
+      const all: Array<Observable<DataQueryResponse>> = [];
+      for (let target of targets) {
+        if (target.hide) {
+          continue;
+        }
+        all.push(this.doSingle(target, request));
+      }
+      return merge(...all);
+    } else {
+      return super.query(request);
+    }
+  }
+
+  storeQuery(target: RedshiftQuery, queryInfo: RedshiftRunningQueryInfo) {
+    const key = JSON.stringify(target);
+    const existingQueryInfo = this.runningQueries[key] || {};
+    this.runningQueries[key] = { ...existingQueryInfo, ...queryInfo };
+  }
+
+  getQuery(target: RedshiftQuery) {
+    const key = JSON.stringify(target);
+    return this.runningQueries[key] || {};
+  }
+
+  removeQuery(target: RedshiftQuery) {
+    const key = JSON.stringify(target);
+    delete this.runningQueries[key];
+  }
+
+  doSingle(target: RedshiftQuery, request: DataQueryRequest<RedshiftQuery>): Observable<DataQueryResponse> {
+    let queryID: string | undefined = undefined;
+    let status: string | undefined = undefined;
+    let allData: DataFrame[] = [];
+
+    return getRequestLooper(
+      { ...request, targets: [target], requestId: `aws_ts_${requestCounter++}` },
+      {
+        /**
+         * Additional query to execute if the current query is still in a running state
+         */
+        getNextQuery: (rsp: DataQueryResponse) => {
+          if (rsp.data?.length) {
+            const first: DataFrame = rsp.data[0];
+            const meta = first.meta?.custom;
+
+            if (isRedshiftCustomMeta(meta) && isRunning(meta.status)) {
+              queryID = meta.queryID;
+              status = meta.status;
+              this.storeQuery(target, { queryID });
+              return { ...target, queryID };
+            }
+          }
+
+          this.removeQuery(target);
+          return undefined;
+        },
+
+        /**
+         * The original request
+         */
+        query: (request: DataQueryRequest<RedshiftQuery>) => {
+          const { range, targets, requestId, intervalMs, maxDataPoints } = request;
+          const [_query] = targets;
+          const query: RedshiftQuery = {
+            ..._query,
+            ...(config.featureToggles.redshiftAsyncQueryDataSupport ? { meta: { queryFlow: 'async' } } : {}),
+          };
+
+          const data = {
+            queries: [
+              {
+                ...query,
+                intervalMs,
+                maxDataPoints,
+                datasource: this?.getRef(),
+                datasourceId: this.id,
+                ...this.applyTemplateVariables(query, request.scopedVars),
+              },
+            ],
+            range: range,
+            from: range.from.valueOf().toString(),
+            to: range.to.valueOf().toString(),
+          };
+
+          let headers = {};
+          if (isRunning(status)) {
+            headers = { 'X-Cache-Skip': true };
```

Comment:
This is for Grafana Enterprise if query caching is enabled.

When we run a query, the backend will respond with a query id. The frontend then tries to get the results and running status of the query from the backend by using this query id. If the backend responds with the status of the query being in a `running` state, that response is cached and we would just keep getting that response until the cache expires. By setting this header, we can bypass the cache only if the query is still running.

Reply:
again lets put this kind of comment in the codebase